### PR TITLE
Avoid rsync of .git directory

### LINKED ISSUE #50272: build/run.sh rsyncs .git directory, causing massive transfers
/kind bug

**What happened**:
After #46193 k8s master started rsyncing `.git` directory, causing hundreds of megabytes to be transferred, making some workflows that utilize remote docker engines impractical. E.g. I use [kubeadm-dind-cluster](https://github.com/Mirantis/kubeadm-dind-cluster) which can start docker-in-docker test cluster on the same remote docker where k8s is built, and currently my Internet connection can only give me about 1 Mb/s, so ~430Mb transfer caused by .git is impractical.

**What you expected to happen**:
`.git` is ignored as before, unless it's needed by the current script. In case `.git` is about to be `rsync`ed, a warning needs to be printed.

**How to reproduce it (as minimally and precisely as possible)**:
Just run `build/run.sh`. Doing so against a remote docker over slow(ish) Internet connection will make the behavior apparent.

**Anything else we need to know?**:
For now perhaps only `update-staging-godeps` script needs to rsync `.git` IIUC (see #46193).

**Environment**:
- Kubernetes version (use `kubectl version`):
654de23cbf821cd8b47387308de9a62aebb6df0d

I'll work on this issue.

### PR DESCRIPTION
**Pull request by dims**
**What this PR does / why we need it**:

We don't really need the .git directory to be transferred. This
problem was introduced in the PR #46193:
"Run the update-staging-godeps script inside a docker container"

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Fixes #50272

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by MHBauer — APPROVED** [review]

**Comment by dims** [comment]
retest

**Comment by dims** [comment]
@ivan4th Can you check if this helps in your case?

**Comment by ivan4th** [comment]
@dims sure, will test it later today, note though that before the change in the script there were extra filters that were preventing generated files from being rsync'd on every `build/run.sh` run -- https://github.com/kubernetes/kubernetes/pull/46193/files#diff-0ee501c7b4b1ab61678579826c6d6534L683
Maybe we should not rsync them unless the script that requires them is run?

**Comment by dims** [comment]
@ivan4th : If those contribute significantly, we can do that in another PR. let's fix the bulk of the stuff due to .git here.

**Comment by ivan4th** [comment]
Actually, these contribute to (re)build time a lot (including even the case when you use Docker on the same machine!) because basically it works like this: generated files are often changed on the build container side during the initial build. When you say change a line somewhere in k8s and run `build/run.sh make WHAT=cmd/kubelet`, your local version of generated files gets rsynced into the build container, thus causing the rebuild of almost everything (as the files are updated). But AFAIK there was a problem with this even without that `.git`-rsyncing PR, so indeed let's check this separately.

**Comment by cblecker** [comment]
/assign
I think I tried this and there was a complication because of it.. let me take a look. If this does work, then I'm all for it.

File: hack/update-staging-godeps-dockerized.sh [hunk 1]
```diff
@@ -87,6 +87,15 @@ function diffGodepManifest() {
   fi
 }
 
+# Create a fake git repo for staging to prevent godeps from complaining
+pushd "${KUBE_ROOT}/staging/src" >/dev/null
```
Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: Is there a reason you're doing this for `${KUBE_ROOT}/staging/src` as opposed to just doing it in `${KUBE_ROOT}`?

File: build/common.sh [hunk 2]
```diff
@@ -682,6 +682,7 @@ function kube::build::sync_to_container() {
   # they will not be re-generated by 'make'.
   kube::build::rsync \
     --delete \
+    --filter='- /.git/' \
```
Comment (cblecker) [inline build/common.sh]: If we're re-adding this, may as well prevent generated files from being rsync'd in too. My original patch:
```
diff --git a/build/common.sh b/build/common.sh
index f6def8f3e3bae..1223510dd0410 100755
--- a/build/common.sh
+++ b/build/common.sh
@@ -682,14 +682,10 @@ function kube::build::sync_to_container() {
   # they will not be re-generated by 'make'.
   kube::build::rsync \
     --delete \
-    --filter='+ /staging/**' \
-    --filter='- /.git/' \
     --filter='- /.make/' \
     --filter='- /_tmp/' \
     --filter='- /_output/' \
     --filter='- /' \
-    --filter='- zz_generated.*' \
-    --filter='- generated.proto' \
     "${KUBE_ROOT}/" "rsync://k8s@${KUBE_RSYNC_ADDR}/k8s/"
```

**Review by cblecker — COMMENTED** [review]
I kicked the tires on this, and this seems to work! Couple questions/suggestions:

File: build/common.sh (hunk 2 above)
Comment (dims) [inline build/common.sh]: @cblecker ack. on it

File: build/common.sh [hunk 3]
```diff
@@ -682,10 +682,14 @@ function kube::build::sync_to_container() {
   # they will not be re-generated by 'make'.
   kube::build::rsync \
     --delete \
+    --filter='+ /staging/**' \
```
Comment (cblecker) [inline build/common.sh]: My bad, my comment wasn't totally clear. This line shouldn't have been restored.. just the generated ones below. Sorry!

File: hack/update-staging-godeps-dockerized.sh (hunk 1 above)
Comment (dims) [inline hack/update-staging-godeps-dockerized.sh]: would be easier to spot that this .git is a fake one

File: build/common.sh (hunk 3 above)
Comment (dims) [inline build/common.sh]: fixed. thanks!

File: hack/update-staging-godeps-dockerized.sh (hunk 1 above)
Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: Hmm.. the filter though is at the root though. I wonder if in some combination, this fake git would be copied out of the container (either via this or another script)

Comment (dims) [inline hack/update-staging-godeps-dockerized.sh]: @cblecker y if we spot such shenanigans we can roll things back

Comment (cblecker) [inline hack/update-staging-godeps-dockerized.sh]: @dims So I was thinking about this solution some more, particularly in the context of changing the `hack/verify-godeps.sh` script too.. we'd need to do this in root as well, so perhaps it's worth being consistent and just run this at `"${KUBE_ROOT}"`? Or would you mind if this was changed to `"${KUBE_ROOT}"` at a later date?

Comment (dims) [inline hack/update-staging-godeps-dockerized.sh]: @cblecker no worries. i've updated the patch

**Pull request merged**